File: code/packages/typescript/ca-capability-analyzer/src/manifest.ts
/**
 * # Capability Manifest — Declaration vs. Reality
 *
 * ## The Core Problem
 *
 * A capability manifest is a JSON file (`required_capabilities.json`) where a
 * project declares what OS capabilities it needs. The manifest says "I need
 * filesystem read access and network connections." The analyzer then checks
 * the actual code to see if reality matches the declaration.
 *
 * There are two kinds of problems we can find:
 *
 * 1. **Undeclared capabilities**: The code uses something not in the manifest.
 *    This is a security concern — the code is doing more than it says it does.
 *
 * 2. **Unused declarations**: The manifest declares a capability the code never
 *    uses. This isn't a security risk, but it's sloppy — like requesting
 *    permissions you don't need.
 *
 * ## Manifest Format
 *
 * The manifest file is a JSON object with this structure:
 *
 * ```json
 * {
 *   "name": "my-package",
 *   "version": "1.0.0",
 *   "capabilities": [
 *     { "category": "fs", "action": "read", "target": "/data/*" },
 *     { "category": "net", "action": "connect", "target": "*" }
 *   ]
 * }
 * ```
 *
 * ## Glob Matching
 *
 * Targets in the manifest support simple glob patterns:
 * - `*` matches anything
 * - `*.txt` matches files ending in .txt
 * - `/data/*` matches anything under /data/
 *
 * We implement a simple glob matcher rather than pulling in a dependency.
 */

import { DetectedCapability } from "./analyzer.js";

// ============================================================================
// Types
// ============================================================================

/**
 * A single capability declaration in the manifest.
 * This represents what the project *says* it needs.
 */
export interface DeclaredCapability {
  /** The broad category: fs, net, proc, env, ffi */
  category: string;
  /** The specific action: read, write, exec, connect, * for any */
  action: string;
  /** The target pattern — may include globs like /data/* */
  target: string;
}

/**
 * The complete manifest file structure.
 */
export interface CapabilityManifest {
  /** Package name */
  name: string;
  /** Package version */
  version: string;
  /** List of declared capabilities */
  capabilities: DeclaredCapability[];
}

/**
 * The result of comparing detected capabilities against the manifest.
 */
export interface ComparisonResult {
  /** Detected capabilities not covered by any manifest entry */
  undeclared: DetectedCapability[];
  /** Manifest entries with no matching detected capability */
  unused: DeclaredCapability[];
  /** Detected capabilities that match a manifest entry */
  matched: DetectedCapability[];
}

// ============================================================================
// Manifest Loading
// ============================================================================

/**
 * ## parseManifest — Parse a manifest from JSON
 *
 * Takes the raw JSON string of a `required_capabilities.json` file and
 * parses it into a structured CapabilityManifest.
 *
 * Performs basic validation:
 * - Must be a valid JSON object
 * - Must have a `capabilities` array
 * - Each capability must have category, action, and target strings
 *
 * @param json - The raw JSON string
 * @returns A parsed CapabilityManifest
 * @throws Error if the JSON is invalid or doesn't match the expected format
 *
 * @example
 * ```ts
 * const manifest = parseManifest(`{
 *   "name": "my-app",
 *   "version": "1.0.0",
 *   "capabilities": [
 *     { "category": "fs", "action": "read", "target": "*" }
 *   ]
 * }`);
 * ```
 */
export function parseManifest(json: string): CapabilityManifest {
  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch {
    throw new Error("Invalid JSON in capability manifest");
  }

  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new Error("Capability manifest must be a JSON object");
  }

  const obj = parsed as Record<string, unknown>;

  if (!Array.isArray(obj.capabilities)) {
    throw new Error("Capability manifest must have a 'capabilities' array");
  }

  const capabilities: DeclaredCapability[] = [];
  for (let i = 0; i < obj.capabilities.length; i++) {
    const cap = obj.capabilities[i];
    if (typeof cap !== "object" || cap === null) {
      throw new Error(`capabilities[${i}] must be an object`);
    }

    const c = cap as Record<string, unknown>;
    if (typeof c.category !== "string") {
      throw new Error(`capabilities[${i}].category must be a string`);
    }
    if (typeof c.action !== "string") {
      throw new Error(`capabilities[${i}].action must be a string`);
    }
    if (typeof c.target !== "string") {
      throw new Error(`capabilities[${i}].target must be a string`);
    }

    capabilities.push({
      category: c.category,
      action: c.action,
      target: c.target,
    });
  }

  return {
    name: typeof obj.name === "string" ? obj.name : "",
    version: typeof obj.version === "string" ? obj.version : "",
    capabilities,
  };
}

// ============================================================================
// Glob Matching
// ============================================================================

/**
 * ## simpleGlobMatch — Minimal Glob Matcher
 *
 * We implement glob matching ourselves to avoid pulling in `minimatch` or
 * similar dependencies. This keeps the analyzer zero-dependency (besides
 * the TypeScript compiler API).
 *
 * ### Supported Patterns
 *
 * | Pattern    | Meaning                        | Example Match         |
 * |------------|--------------------------------|----------------------|
 * | `*`        | Matches anything               | (anything)           |
 * | `*.txt`    | Matches suffix                 | "readme.txt"         |
 * | `/data/*`  | Matches prefix                 | "/data/file.csv"     |
 * | `exact`    | Exact string match             | "exact" only         |
 *
 * ### How it works
 *
 * We convert the glob pattern to a regular expression:
 * - `*` becomes `.*` (match anything)
 * - `.` and other regex metacharacters are escaped
 * - The pattern is anchored to match the full string (^ and $)
 *
 * @param pattern - The glob pattern (from the manifest)
 * @param value - The actual value (from detection)
 * @returns true if the value matches the pattern
 *
 * @example
 * ```ts
 * simpleGlobMatch("*", "anything")           // true
 * simpleGlobMatch("/data/*", "/data/x.csv")  // true
 * simpleGlobMatch("*.txt", "readme.md")      // false
 * ```
 */
export function simpleGlobMatch(pattern: string, value: string): boolean {
  // Fast path: wildcard matches everything
  if (pattern === "*") return true;

  // Fast path: exact match
  if (!pattern.includes("*")) return pattern === value;

  // Convert glob to regex:
  // 1. Escape regex metacharacters (except *)
  // 2. Replace * with .*
  // 3. Anchor with ^ and $
  const escaped = pattern
    .replace(/[.+?^${}()|[\]\\]/g, "\\$&")  // Escape regex specials
    .replace(/\*/g, ".*");                    // Convert * to .*

  const regex = new RegExp(`^${escaped}$`);
  return regex.test(value);
}

// ============================================================================
// Capability Comparison
// ============================================================================

/**
 * ## capabilityMatchesDeclaration — Single Capability Check
 *
 * Determines whether a single detected capability is covered by a single
 * manifest declaration.
 *
 * ### Matching Rules
 *
 * For a detected capability to match a declaration:
 * 1. **Category must match exactly** — fs != net
 * 2. **Action must match** — either exactly, or the declaration uses "*"
 * 3. **Target must match** — either exactly, via glob, or the declaration uses "*"
 *
 * ### Truth Table
 *
 * | Declared         | Detected            | Match? | Why                        |
 * |------------------|---------------------|--------|----------------------------|
 * | fs:read:*        | fs:read:/etc/passwd | Yes    | * target matches anything  |
 * | fs:*:*           | fs:write:/tmp/x     | Yes    | * action and target        |
 * | fs:read:*.txt    | fs:read:data.txt    | Yes    | Glob match on target       |
 * | fs:read:*        | fs:write:/tmp/x     | No     | Action mismatch            |
 * | net:connect:*    | fs:read:/etc/passwd | No     | Category mismatch          |
 */
export function capabilityMatchesDeclaration(
  detected: DetectedCapability,
  declared: DeclaredCapability,
): boolean {
  // Category must match exactly
  if (detected.category !== declared.category) return false;

  // Action must match (declared * matches any action)
  if (declared.action !== "*" && detected.action !== declared.action) return false;

  // Target must match via glob (declared * matches any target)
  if (!simpleGlobMatch(declared.target, detected.target)) return false;

  return true;
}

/**
 * ## compareCapabilities — The Main Comparison Function
 *
 * Compares all detected capabilities against all manifest declarations.
 * Produces three lists:
 *
 * 1. **matched**: Detected capabilities covered by the manifest (good)
 * 2. **undeclared**: Detected capabilities NOT in the manifest (security concern)
 * 3. **unused**: Manifest declarations with no matching detection (sloppy)
 *
 * ### Algorithm
 *
 * For each detected capability, we scan all manifest declarations looking
 * for a match. If at least one declaration matches, the capability is "covered."
 * If none match, it's "undeclared."
 *
 * Then, for each manifest declaration, we check if any detected capability
 * matched it. If none did, it's "unused."
 *
 * This is O(n*m) where n is detected capabilities and m is declarations.
 * For typical projects (tens to hundreds of capabilities, single-digit
 * declarations), this is perfectly fast.
 *
 * @param detected - All capabilities found by the analyzer
 * @param manifest - The parsed capability manifest
 * @returns A ComparisonResult with matched, undeclared, and unused lists
 *
 * @example
 * ```ts
 * const result = compareCapabilities(
 *   [{ category: "fs", action: "read", target: "/etc/passwd", ... }],
 *   { capabilities: [{ category: "fs", action: "read", target: "*" }] }
 * );
 * // result.matched.length === 1, result.undeclared.length === 0
 * ```
 */
export function compareCapabilities(
  detected: DetectedCapability[],
  manifest: CapabilityManifest,
): ComparisonResult {
  const matched: DetectedCapability[] = [];
  const undeclared: DetectedCapability[] = [];

  /**
   * Track which manifest declarations were matched.
   * We use an array of booleans parallel to manifest.capabilities.
   */
  const declarationUsed = new Array(manifest.capabilities.length).fill(false) as boolean[];

  for (const cap of detected) {
    let isMatched = false;

    for (let i = 0; i < manifest.capabilities.length; i++) {
      if (capabilityMatchesDeclaration(cap, manifest.capabilities[i])) {
        isMatched = true;
        declarationUsed[i] = true;
        // Don't break — a capability could match multiple declarations,
        // and we want to mark all matching declarations as "used"
      }
    }

    if (isMatched) {
      matched.push(cap);
    } else {
      undeclared.push(cap);
    }
  }

  // Find unused declarations
  const unused: DeclaredCapability[] = manifest.capabilities
    .filter((_, i) => !declarationUsed[i]);

  return { matched, undeclared, unused };
}

File: code/packages/rust/ca-capability-analyzer/src/manifest.rs
//! # Manifest Loading and Capability Comparison
//!
//! This module loads a package's `required_capabilities.json` manifest and
//! compares it against the capabilities detected by the analyzer. The
//! comparison answers the question: *"Does this package use only the
//! capabilities it declared?"*
//!
//! ## Comparison Logic
//!
//! The comparison is **asymmetric**:
//!
//! - **Undeclared capabilities** (detected but not in manifest) are **ERRORS**.
//!   The code uses something it didn't declare. This is a security violation.
//!
//! - **Unused declarations** (in manifest but not detected) are **WARNINGS**.
//!   The manifest declares a capability the code doesn't use. This isn't
//!   a security issue — it's just a stale declaration that should be cleaned up.
//!
//! ```text
//! ┌────────────────────┐    ┌────────────────────┐
//! │  Detected by       │    │  Declared in        │
//! │  Analyzer          │    │  Manifest           │
//! │                    │    │                     │
//! │  ┌──────────┐      │    │  ┌──────────┐       │
//! │  │ MATCHED  │◄─────┼────┼──│ MATCHED  │       │
//! │  └──────────┘      │    │  └──────────┘       │
//! │  ┌──────────┐      │    │  ┌──────────┐       │
//! │  │ ERROR    │      │    │  │ WARNING  │       │
//! │  │(undeclrd)│      │    │  │ (unused) │       │
//! │  └──────────┘      │    │  └──────────┘       │
//! └────────────────────┘    └────────────────────┘
//! ```
//!
//! ## Default Deny
//!
//! If no `required_capabilities.json` exists, the package is treated as
//! having **zero** declared capabilities. Any detected capability is an
//! error. This is the "no manifest = block everything" principle.
//!
//! ## Target Matching
//!
//! When comparing detected targets against declared targets, we use
//! **fnmatch-style glob matching**:
//!
//! - `*` matches anything
//! - `data/*.txt` matches `data/config.txt`
//! - `config.toml` matches `config.toml` exactly
//!
//! This mirrors OpenBSD's `unveil()` path matching behavior.

use serde::{Deserialize, Serialize};

use crate::analyzer::DetectedCapability;

// ── Data Structures ──────────────────────────────────────────────────

/// A declared capability in the manifest.
///
/// Each capability has three fields that define what the package is
/// allowed to do:
///
/// - `category` — The resource kind (fs, net, proc, env, ffi)
/// - `action` — The operation (read, write, connect, exec, *, etc.)
/// - `target` — The specific resource or glob pattern ("*", "data/*.txt")
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeclaredCapability {
    pub category: String,
    pub action: String,
    pub target: String,
}

/// A parsed `required_capabilities.json` manifest.
///
/// This struct mirrors the JSON structure of the manifest file:
///
/// ```json
/// {
///   "package": "rust/my-package",
///   "capabilities": [
///     { "category": "fs", "action": "read", "target": "*.toml" }
///   ],
///   "justification": "Reads config files",
///   "banned_construct_exceptions": []
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub package: String,
    pub capabilities: Vec<DeclaredCapability>,
    #[serde(default)]
    pub justification: String,
    #[serde(default)]
    pub banned_construct_exceptions: Vec<serde_json::Value>,
    /// Path to the manifest file (not serialized from JSON)
    #[serde(skip)]
    pub path: Option<String>,
}

impl Manifest {
    /// True if the manifest declares zero capabilities.
    pub fn is_empty(&self) -> bool {
        self.capabilities.is_empty()
    }
}

/// Load a manifest from a JSON file.
///
/// # Arguments
///
/// * `path` — Path to `required_capabilities.json`
///
/// # Returns
///
/// A parsed `Manifest`, or an error string.
///
/// # Errors
///
/// Returns an error if the file can't be read or isn't valid JSON.
pub fn load_manifest(path: &str) -> Result<Manifest, String> {
    let content = std::fs::read_to_string(path)
        .map_err(|e| format!("Failed to read manifest {}: {}", path, e))?;

    let mut manifest: Manifest = serde_json::from_str(&content)
        .map_err(|e| format!("Failed to parse manifest {}: {}", path, e))?;

    manifest.path = Some(path.to_string());
    Ok(manifest)
}

/// Create a default (empty) manifest for a package without one.
///
/// This represents the **default deny** policy: a package without a
/// `required_capabilities.json` is treated as declaring zero capabilities.
/// Any detected capability will be flagged as an error.
pub fn default_manifest(package_name: &str) -> Manifest {
    Manifest {
        package: package_name.to_string(),
        capabilities: Vec::new(),
        justification: "No manifest file — default deny (zero capabilities).".to_string(),
        banned_construct_exceptions: Vec::new(),
        path: None,
    }
}

// ── Comparison Result ────────────────────────────────────────────────

/// Result of comparing detected capabilities against a manifest.
///
/// This is the output of the core comparison logic. The CI gate uses
/// `passed` to determine whether to allow or block a package.
#[derive(Debug)]
pub struct ComparisonResult {
    /// True if all detected capabilities are covered by declarations.
    pub passed: bool,
    /// Detected capabilities not covered by any declaration (violations).
    pub errors: Vec<DetectedCapability>,
    /// Declared capabilities not matched by any detection (stale).
    pub warnings: Vec<DeclaredCapability>,
    /// Detected capabilities that matched a declaration.
    pub matched: Vec<DetectedCapability>,
}

impl ComparisonResult {
    /// Human-readable summary of the comparison result.
    pub fn summary(&self) -> String {
        let mut lines = Vec::new();

        if self.passed {
            lines.push("PASS — all detected capabilities are declared.".to_string());
        } else {
            lines.push(format!(
                "FAIL — {} undeclared capability(ies) detected.",
                self.errors.len()
            ));
        }

        if !self.errors.is_empty() {
            lines.push(String::new());
            lines.push("Undeclared capabilities (ERRORS):".to_string());
            for cap in &self.errors {
                lines.push(format!(
                    "  {}:{}: {} ({})",
                    cap.file, cap.line, cap, cap.evidence
                ));
            }
        }

        if !self.warnings.is_empty() {
            lines.push(String::new());
            lines.push("Unused declarations (WARNINGS):".to_string());
            for decl in &self.warnings {
                lines.push(format!(
                    "  {}:{}:{}",
                    decl.category, decl.action, decl.target
                ));
            }
        }

        if !self.matched.is_empty() {
            lines.push(format!("\nMatched: {} capability(ies).", self.matched.len()));
        }

        lines.join("\n")
    }
}

// ── Target Matching (fnmatch-style glob) ─────────────────────────────
//
// We implement a simple fnmatch-style glob matcher. The key patterns:
//
// - `*` matches any sequence of characters (but not path separators
//   in strict mode — we use loose mode here for simplicity)
// - `?` matches any single character
// - Literal characters match themselves
//
// We don't use an external crate to keep dependencies minimal.

/// Check if a string matches an fnmatch-style glob pattern.
///
/// Supports `*` (match any sequence) and `?` (match single char).
///
/// # Examples
///
/// ```text
/// fnmatch("*", "anything")          → true
/// fnmatch("*.txt", "data.txt")      → true
/// fnmatch("data/?.txt", "data/a.txt") → true
/// fnmatch("exact", "exact")         → true
/// fnmatch("exact", "other")         → false
/// ```
pub fn fnmatch(pattern: &str, text: &str) -> bool {
    // Convert pattern and text to char vectors for indexed access
    let pattern: Vec<char> = pattern.chars().collect();
    let text: Vec<char> = text.chars().collect();

    // Dynamic programming approach:
    // dp[i][j] = true if pattern[0..i] matches text[0..j]
    //
    // This is a classic algorithm — each cell depends on:
    // - If pattern[i-1] == '*': dp[i][j] = dp[i-1][j] (skip *) OR dp[i][j-1] (consume char)
    // - If pattern[i-1] == '?' or matches text[j-1]: dp[i][j] = dp[i-1][j-1]
    // - Otherwise: dp[i][j] = false

    let p_len = pattern.len();
    let t_len = text.len();

    let mut dp = vec![vec![false; t_len + 1]; p_len + 1];
    dp[0][0] = true; // Empty pattern matches empty text

    // Handle leading '*' patterns that match empty text
    for i in 1..=p_len {
        if pattern[i - 1] == '*' {
            dp[i][0] = dp[i - 1][0];
        }
    }

    for i in 1..=p_len {
        for j in 1..=t_len {
            if pattern[i - 1] == '*' {
                // '*' can match zero chars (dp[i-1][j]) or consume one more char (dp[i][j-1])
                dp[i][j] = dp[i - 1][j] || dp[i][j - 1];
            } else if pattern[i - 1] == '?' || pattern[i - 1] == text[j - 1] {
                dp[i][j] = dp[i - 1][j - 1];
            }
        }
    }

    dp[p_len][t_len]
}

/// Check if a detected target matches a declared target pattern.
///
/// This wraps `fnmatch` with two special cases:
///
/// 1. If the declared pattern is `"*"`, it matches anything.
/// 2. If the detected target is `"*"` (unknown/wildcard), it matches
///    any declared pattern. This is conservative: we accept it rather
///    than generating a false positive, since we can't determine what
///    the actual runtime target will be.
fn target_matches(pattern: &str, actual: &str) -> bool {
    if pattern == "*" {
        return true;
    }
    if actual == "*" {
        // The detected target is a wildcard (non-literal). We can't know
        // what it will resolve to at runtime, so we accept it as matching
        // any declared pattern. This avoids false positives.
        return true;
    }
    fnmatch(pattern, actual)
}

/// Check if a detected capability matches a declared capability.
///
/// A match requires all three components to be compatible:
///
/// 1. **Category** must match exactly (fs == fs, but fs != net)
/// 2. **Action** must match exactly, or the declared action is `"*"`
/// 3. **Target** must match via glob (fnmatch)
fn capability_matches(declared: &DeclaredCapability, detected: &DetectedCapability) -> bool {
    if declared.category != detected.category {
        return false;
    }
    if declared.action != "*" && declared.action != detected.action {
        return false;
    }
    target_matches(&declared.target, &detected.target)
}

// ── Core Comparison Logic ────────────────────────────────────────────

/// Compare detected capabilities against a manifest.
///
/// This is the core comparison logic used by the CI gate. It determines
/// whether a package's source code uses only the capabilities it declared.
///
/// # Arguments
///
/// * `detected` — Capabilities found by the analyzer
/// * `manifest` — The package's declared capabilities
///
/// # Returns
///
/// A `ComparisonResult` with pass/fail status, errors, and warnings.
pub fn compare_capabilities(
    detected: &[DetectedCapability],
    manifest: &Manifest,
) -> ComparisonResult {
    let mut errors = Vec::new();
    let mut matched = Vec::new();

    // For each detected capability, check if any declaration covers it
    for cap in detected {
        let found_match = manifest
            .capabilities
            .iter()
            .any(|decl| capability_matches(decl, cap));

        if found_match {
            matched.push(cap.clone());
        } else {
            errors.push(cap.clone());
        }
    }

    // Find unused declarations (warnings)
    let warnings: Vec<DeclaredCapability> = manifest
        .capabilities
        .iter()
        .filter(|decl| !detected.iter().any(|cap| capability_matches(decl, cap)))
        .cloned()
        .collect();

    ComparisonResult {
        passed: errors.is_empty(),
        errors,
        warnings,
        matched,
    }
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── Helper: create a DetectedCapability quickly ──

    fn det(category: &str, action: &str, target: &str) -> DetectedCapability {
        DetectedCapability {
            category: category.to_string(),
            action: action.to_string(),
            target: target.to_string(),
            file: "test.rs".to_string(),
            line: 1,
            evidence: "test".to_string(),
        }
    }

    fn decl(category: &str, action: &str, target: &str) -> DeclaredCapability {
        DeclaredCapability {
            category: category.to_string(),
            action: action.to_string(),
            target: target.to_string(),
        }
    }

    fn manifest_with(caps: Vec<DeclaredCapability>) -> Manifest {
        Manifest {
            package: "test/pkg".to_string(),
            capabilities: caps,
            justification: "test".to_string(),
            banned_construct_exceptions: Vec::new(),
            path: None,
        }
    }

    // ── fnmatch tests ────────────────────────────────────────────────

    #[test]
    fn test_fnmatch_exact() {
        assert!(fnmatch("hello", "hello"));
        assert!(!fnmatch("hello", "world"));
    }

    #[test]
    fn test_fnmatch_star_matches_anything() {
        assert!(fnmatch("*", "anything"));
        assert!(fnmatch("*", ""));
        assert!(fnmatch("*", "multi/path/here"));
    }

    #[test]
    fn test_fnmatch_star_prefix() {
        assert!(fnmatch("*.txt", "data.txt"));
        assert!(fnmatch("*.txt", ".txt"));
        assert!(!fnmatch("*.txt", "data.csv"));
    }

    #[test]
    fn test_fnmatch_star_suffix() {
        assert!(fnmatch("data.*", "data.txt"));
        assert!(fnmatch("data.*", "data."));
        assert!(!fnmatch("data.*", "info.txt"));
    }

    #[test]
    fn test_fnmatch_star_middle() {
        assert!(fnmatch("data/*.txt", "data/config.txt"));
        assert!(fnmatch("data/*.txt", "data/x.txt"));
        assert!(!fnmatch("data/*.txt", "other/config.txt"));
    }

    #[test]
    fn test_fnmatch_question_mark() {
        assert!(fnmatch("?.txt", "a.txt"));
        assert!(!fnmatch("?.txt", "ab.txt"));
        assert!(!fnmatch("?.txt", ".txt"));
    }

    #[test]
    fn test_fnmatch_complex_pattern() {
        assert!(fnmatch("../../grammars/*.tokens", "../../grammars/python.tokens"));
        assert!(!fnmatch("../../grammars/*.tokens", "../../grammars/python.rules"));
    }

    #[test]
    fn test_fnmatch_empty() {
        assert!(fnmatch("", ""));
        assert!(!fnmatch("", "a"));
        assert!(!fnmatch("a", ""));
    }

    #[test]
    fn test_fnmatch_multiple_stars() {
        assert!(fnmatch("*/*", "a/b"));
        assert!(fnmatch("**", "anything"));
    }

    // ── target_matches tests ─────────────────────────────────────────

    #[test]
    fn test_target_matches_wildcard_pattern() {
        assert!(target_matches("*", "anything"));
        assert!(target_matches("*", ""));
    }

    #[test]
    fn test_target_matches_wildcard_actual() {
        // When detected target is "*", we accept any declared pattern
        assert!(target_matches("specific.txt", "*"));
        assert!(target_matches("*.txt", "*"));
    }

    #[test]
    fn test_target_matches_exact() {
        assert!(target_matches("data.txt", "data.txt"));
        assert!(!target_matches("data.txt", "other.txt"));
    }

    #[test]
    fn test_target_matches_glob() {
        assert!(target_matches("*.toml", "config.toml"));
        assert!(!target_matches("*.toml", "config.json"));
    }

    // ── capability_matches tests ─────────────────────────────────────

    #[test]
    fn test_capability_matches_exact() {
        let declared = decl("fs", "read", "data.txt");
        let detected = det("fs", "read", "data.txt");
        assert!(capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_wrong_category() {
        let declared = decl("fs", "read", "*");
        let detected = det("net", "read", "*");
        assert!(!capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_wrong_action() {
        let declared = decl("fs", "read", "*");
        let detected = det("fs", "write", "*");
        assert!(!capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_wildcard_action() {
        let declared = decl("fs", "*", "*");
        let detected = det("fs", "read", "data.txt");
        assert!(capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_wildcard_target() {
        let declared = decl("fs", "read", "*");
        let detected = det("fs", "read", "data.txt");
        assert!(capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_glob_target() {
        let declared = decl("fs", "read", "*.toml");
        let detected = det("fs", "read", "config.toml");
        assert!(capability_matches(&declared, &detected));
    }

    #[test]
    fn test_capability_matches_glob_target_no_match() {
        let declared = decl("fs", "read", "*.toml");
        let detected = det("fs", "read", "config.json");
        assert!(!capability_matches(&declared, &detected));
    }

    // ── compare_capabilities tests ───────────────────────────────────

    #[test]
    fn test_compare_all_matched() {
        let detected = vec![det("fs", "read", "data.txt")];
        let manifest = manifest_with(vec![decl("fs", "read", "*")]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(result.passed);
        assert!(result.errors.is_empty());
        assert_eq!(result.matched.len(), 1);
    }

    #[test]
    fn test_compare_undeclared_error() {
        let detected = vec![det("net", "connect", "evil.com:443")];
        let manifest = manifest_with(vec![decl("fs", "read", "*")]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(!result.passed);
        assert_eq!(result.errors.len(), 1);
        assert_eq!(result.errors[0].category, "net");
    }

    #[test]
    fn test_compare_unused_warning() {
        let detected = vec![det("fs", "read", "data.txt")];
        let manifest = manifest_with(vec![
            decl("fs", "read", "*"),
            decl("net", "connect", "*"),
        ]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(result.passed); // Unused declarations don't fail
        assert_eq!(result.warnings.len(), 1);
        assert_eq!(result.warnings[0].category, "net");
    }

    #[test]
    fn test_compare_empty_manifest_default_deny() {
        let detected = vec![det("fs", "read", "data.txt")];
        let manifest = default_manifest("test/pkg");

        let result = compare_capabilities(&detected, &manifest);
        assert!(!result.passed);
        assert_eq!(result.errors.len(), 1);
    }

    #[test]
    fn test_compare_no_detections_passes() {
        let detected: Vec<DetectedCapability> = vec![];
        let manifest = manifest_with(vec![decl("fs", "read", "*")]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(result.passed);
        assert!(result.errors.is_empty());
        // The declaration is unused
        assert_eq!(result.warnings.len(), 1);
    }

    #[test]
    fn test_compare_multiple_errors() {
        let detected = vec![
            det("fs", "read", "data.txt"),
            det("net", "connect", "api.com:443"),
            det("proc", "exec", "ls"),
        ];
        let manifest = manifest_with(vec![decl("fs", "read", "*")]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(!result.passed);
        assert_eq!(result.errors.len(), 2); // net and proc are undeclared
        assert_eq!(result.matched.len(), 1); // fs is matched
    }

    #[test]
    fn test_compare_wildcard_action_covers_all() {
        let detected = vec![
            det("fs", "read", "a.txt"),
            det("fs", "write", "b.txt"),
            det("fs", "delete", "c.txt"),
        ];
        let manifest = manifest_with(vec![decl("fs", "*", "*")]);

        let result = compare_capabilities(&detected, &manifest);
        assert!(result.passed);
        assert_eq!(result.matched.len(), 3);
    }

    // ── Manifest loading tests ───────────────────────────────────────

    #[test]
    fn test_default_manifest_is_empty() {
        let m = default_manifest("test/pkg");
        assert!(m.is_empty());
        assert_eq!(m.package, "test/pkg");
    }

    #[test]
    fn test_manifest_is_empty() {
        let m = manifest_with(vec![]);
        assert!(m.is_empty());

        let m2 = manifest_with(vec![decl("fs", "read", "*")]);
        assert!(!m2.is_empty());
    }

    #[test]
    fn test_load_manifest_from_string() {
        let json = r#"{
            "package": "rust/my-pkg",
            "capabilities": [
                { "category": "fs", "action": "read", "target": "*.toml" }
            ],
            "justification": "Reads config"
        }"#;

        let manifest: Manifest = serde_json::from_str(json).unwrap();
        assert_eq!(manifest.package, "rust/my-pkg");
        assert_eq!(manifest.capabilities.len(), 1);
        assert_eq!(manifest.capabilities[0].category, "fs");
        assert_eq!(manifest.capabilities[0].action, "read");
        assert_eq!(manifest.capabilities[0].target, "*.toml");
    }

    #[test]
    fn test_load_manifest_missing_optional_fields() {
        let json = r#"{
            "package": "rust/my-pkg",
            "capabilities": []
        }"#;

        let manifest: Manifest = serde_json::from_str(json).unwrap();
        assert_eq!(manifest.justification, "");
        assert!(manifest.banned_construct_exceptions.is_empty());
    }

    // ── Summary output tests ─────────────────────────────────────────

    #[test]
    fn test_summary_pass() {
        let result = ComparisonResult {
            passed: true,
            errors: vec![],
            warnings: vec![],
            matched: vec![det("fs", "read", "x")],
        };
        let summary = result.summary();
        assert!(summary.contains("PASS"));
    }

    #[test]
    fn test_summary_fail() {
        let result = ComparisonResult {
            passed: false,
            errors: vec![det("net", "connect", "evil.com")],
            warnings: vec![],
            matched: vec![],
        };
        let summary = result.summary();
        assert!(summary.contains("FAIL"));
        assert!(summary.contains("net:connect:evil.com"));
    }

    #[test]
    fn test_summary_with_warnings() {
        let result = ComparisonResult {
            passed: true,
            errors: vec![],
            warnings: vec![decl("proc", "exec", "*")],
            matched: vec![det("fs", "read", "x")],
        };
        let summary = result.summary();
        assert!(summary.contains("PASS"));
        assert!(summary.contains("Unused declarations"));
        assert!(summary.contains("proc:exec:*"));
    }
}
